source/agora/test/Base.d

```diff
@@ -1389,117 +1376,111 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
+
+        return conf;
+    }
 
+    Config makeFullNodeConfig (size_t idx, Address self_address,
+        Address[] addresses)
+    {
         Config conf =
         {
             banman : ban_conf,
-            node : self,
-            validator : validator_self,
-            network : assumeUnique(network),
+            node : makeNodeConfig(self_address),
+            network : makeNetwork(idx, addresses),
         };
 
         return conf;
     }
 
-    auto key_range = WK.Keys.byRange().takeExactly(
-        test_conf.validators + test_conf.outsider_validators).array();
-    auto validator_keys = key_range[0 .. test_conf.validators];
-    key_range.popFrontN(test_conf.validators);
-    auto outsider_validator_keys = key_range[0 .. test_conf.outsider_validators];
-    key_range.popFrontN(test_conf.outsider_validators);
-
-    ValidatorConfig[] validator_configs;
-    NodeConfig[] node_configs;
-    Config[] main_configs;
-
-    validator_keys
-        .each!(key => validator_configs ~= makeValidatorConfig(key));
-    validator_keys
-        .each!(key => node_configs ~= makeNodeConfig(validatorAddress(key)));
-    iota(test_conf.full_nodes)
-        .each!(_ => node_configs ~= makeNodeConfig(fullNodeAddress()));
-
-    // network these nodes together using the configured NetworkTopology
-    node_configs.take(test_conf.validators).zip(validator_keys).enumerate
-        .each!(pair => main_configs ~= makeMainConfig(
-            pair.index, pair.value[0], node_configs,
-                makeValidatorConfig(pair.value[1])));
-    node_configs.drop(test_conf.validators).enumerate(test_conf.validators)
-        .each!(pair => main_configs ~= makeMainConfig(
-            pair.index, pair.value, node_configs));
-
-    NodeConfig[] extra_validator_configs;
-    NodeConfig[] extra_node_configs;
-
-    outsider_validator_keys
-        .each!(key => extra_validator_configs ~= makeNodeConfig(validatorAddress(key)));
-    iota(test_conf.outsider_full_nodes)
-        .each!(_ => extra_node_configs ~= makeNodeConfig(fullNodeAddress()));
-
-    // generate the outsider main_configs. pair.index is correct here despite
-    // passing 'node_configs'. in 'MinimallyConnected' mode the connection will be
-    // n2 <- n1 <- n2 and n1 <- o1 and n2 <- o2, (o = outsider).
-    // otherwise it will be [n1 <-> n2] <- o1 and [n1 <-> n2] <- o2
-
-    extra_validator_configs.zip(outsider_validator_keys).enumerate
-        .each!(pair => main_configs ~= makeMainConfig(
-            pair.index % node_configs.length, pair.value[0], node_configs,
-                makeValidatorConfig(pair.value[1])));
-    extra_node_configs.enumerate(test_conf.outsider_validators).each!(
-        pair => main_configs ~= makeMainConfig(
-            pair.index % node_configs.length, pair.value, node_configs));
+    alias validatorAddress = (idx, key) => format("Validator #%s (%s)", idx,
+        key.address);
+
+    alias fullNodeAddress = (idx) => format("FullNode #%s", idx);
```

Comment:
Ergh, to me it's clearer to have it as a nested function. This is a template literal, but it doesn't need to be templated.

Reply:
okay _sad noises_